pallets/hyperdrive/src/tezos.rs

```diff
@@ -0,0 +1,641 @@
+use sp_core::ConstU32;
+use sp_runtime::traits::Member;
+use sp_std::prelude::*;
+use sp_std::str::FromStr;
+use sp_std::vec;
+use tezos_core::types::encoded::Address as TezosAddress;
+use tezos_core::Error as TezosCoreError;
+use tezos_michelson::micheline::primitive_application::PrimitiveApplication;
+use tezos_michelson::michelson::data;
+use tezos_michelson::michelson::data::{
+    try_int, try_string, Bytes, Data, Int, Nat, Pair, Sequence,
+};
+use tezos_michelson::michelson::types::{
+    address, bool as bool_type, bytes, nat, option, pair, set, string,
+};
+use tezos_michelson::Error as TezosMichelineError;
+use tezos_michelson::{
+    micheline::{primitive_application, Micheline},
+    michelson::ComparableTypePrimitive,
+};
+
+use pallet_acurast::{JobIdSequence, JobRegistration, MultiOrigin, Schedule};
+use pallet_acurast_marketplace::{
+    JobRequirements, MultiDestination, PlannedExecution, RegistrationExtra,
+};
+
+use crate::types::{MessageParser, RawAction};
+use crate::RewardParser;
+use crate::{MessageIdentifier, ParsedAction};
+
+pub struct TezosParser<Reward, Balance, ParsableAccountId, AccountId, Extra, AssetParser>(
+    PhantomData<(
+        Reward,
+        Balance,
+        ParsableAccountId,
+        AccountId,
+        Extra,
+        AssetParser,
+    )>,
+);
+
+impl<Reward, Balance, ParsableAccountId, AccountId, Extra, AssetParser>
+    MessageParser<Reward, AccountId, Extra>
+    for TezosParser<Reward, Balance, ParsableAccountId, AccountId, Extra, AssetParser>
+where
+    ParsableAccountId: FromStr + Into<AccountId>,
+    Extra: From<RegistrationExtra<Reward, Balance, AccountId>>,
+    Reward: Parameter + Member,
+    Balance: From<u128>,
+    AssetParser: RewardParser<Reward>,
+{
+    type Error = ValidationError;
+    type AssetParser = AssetParser;
+
+    fn parse_key(encoded: &[u8]) -> Result<MessageIdentifier, Self::Error> {
+        unpack_nat(encoded)
+    }
+
+    fn parse_value(encoded: &[u8]) -> Result<ParsedAction<AccountId, Extra>, ValidationError> {
+        let (action, origin, payload) = parse_message(encoded)?;
+
+        Ok(match action {
+            RawAction::RegisterJob => {
+                let payload: Vec<u8> = (&payload).into();
+                let (job_id_sequence, registration) = parse_job_registration_payload::<
+                    Reward,
+                    Balance,
+                    ParsableAccountId,
+                    AccountId,
+                    Extra,
+                    AssetParser,
+                >(payload.as_slice())?;
+
+                ParsedAction::RegisterJob(
+                    (
+                        MultiOrigin::Tezos(bounded_address(&origin)?),
+                        job_id_sequence,
+                    ),
+                    registration,
+                )
+            }
+        })
+    }
+}
+
+/// Parses an encoded key from Tezos representing a message identifier.
+pub fn unpack_nat(encoded: &[u8]) -> Result<u128, ValidationError> {
```

Comment:
You could make it generic and name it like that, but the error type returned is specific...so you would have to move to the caller.